texar/modules/pretrained/xlnet_model_utils.py

```diff
@@ -0,0 +1,342 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Model Utils of XLNet Modules.
+
+Adapted from
+https://github.com/zihangdai/xlnet/blob/master/modeling.py
+"""
+
+from typing import Any, Dict, Optional, Tuple
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from texar.core import get_layer
+from texar.module_base import ModuleBase
+
+__all__ = [
+    "PositionWiseFF",
+    "RelativeMultiheadAttention",
+    "RelativePositionalEncoding",
+]
+
+
+class PositionWiseFF(ModuleBase):
+
+    def __init__(self, hparams=None):
+        super().__init__(hparams)
+
+        hidden_dim = self._hparams.hidden_dim
+        ffn_inner_dim = self._hparams.ffn_inner_dim
+        dropout = self._hparams.dropout
+        activation = self._hparams.activation.capitalize()
+        if activation == 'Relu':
+            activation = 'ReLU'
+        elif activation == 'Gelu':
+            activation = 'GPTGELU'
+
+        self.linear1 = nn.Linear(hidden_dim, ffn_inner_dim)
+        self.activation_fn = get_layer({"type": activation})
+        self.dropout = nn.Dropout(dropout, inplace=True)
+        self.linear2 = nn.Linear(ffn_inner_dim, hidden_dim)
+        self.layer_norm = nn.LayerNorm(hidden_dim, eps=1e-12)
+
+    @staticmethod
+    def default_hparams() -> Dict[str, Any]:
+        return {
+            "hidden_dim": 768,
+            "ffn_inner_dim": 3072,
+            "dropout": 0.1,
+            "activation": 'relu',
+        }
+
+    def forward(self,  # type: ignore
+                input: torch.Tensor) -> torch.Tensor:
+        # position-wise feed-forward
+        output = self.linear1(input)
+        output = self.activation_fn(output)
+        output = self.dropout(output)
+        output = self.linear2(output)
+        output = self.dropout(output)
+        # residual + layer norm
+        output = self.layer_norm(input + output)
+        return output
+
+
+class PositionalEmbedding(nn.Module):
+
+    def __init__(self, embed_dim: int):
+        super().__init__()
+
+        freq_seq = torch.arange(0.0, embed_dim, 2.0)
+        inv_freq = 1 / (10000 ** (freq_seq / embed_dim))
+        self.register_buffer('inv_freq', inv_freq)
+
+    def forward(self,  # type: ignore
+                pos_seq: torch.Tensor) -> torch.Tensor:
+        sinusoid = torch.ger(pos_seq, self.inv_freq)
+        pos_embed = torch.cat([sinusoid.sin(), sinusoid.cos()], dim=-1)
+        return pos_embed
+
+
+class RelativePositionalEncoding(ModuleBase):
```

Comment:
Same for this position embedder.

It's possible that these embedders are usable in other (future) models. In the docstring, we can say "this embedder is used in XLNet"

Reply:
TBH this module is specific to XLNet, and all it does is create sinusoidal positional embeddings for specific index ranges. I think we should keep this only if we add `RelativeMultiheadAttention` to Texar modules.